python/ray/data/datasource/datasink.py

```diff
@@ -0,0 +1,77 @@
+from typing import Any, Iterable, List
+
+from ray.data._internal.execution.interfaces import TaskContext
+from ray.data.block import Block
+
+WriteResult = Any
+
+
+class Datasink:
+    def on_write_start(self, **_) -> None:
+        """Callback for when a write job starts.
+
+        Use this method to perform setup for write tasks. For example, creating a
+        staging bucket in S3.
+
+        Args:
+            _: Forward-compatibility placeholder.
+        """
+        pass
+
+    def write(
+        self,
+        blocks: Iterable[Block],
+        ctx: TaskContext,
+        **_,
+    ) -> WriteResult:
+        """Write blocks. This is used by a single write task.
+
+        Args:
+            blocks: List of data blocks.
+            ctx: ``TaskContext`` for the write task.
+            _: Forward-compatibility placeholder.
+
+        Returns:
+            The output of the write task.
```

Comment:
Maybe explain this a bit more, I didn't catch that WriteResult was Any and could be user-defined at first.